src/App.tsx

```diff
@@ -1,26 +1,149 @@
-/* eslint-disable max-len */
+/* eslint-disable jsx-a11y/label-has-associated-control */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
-import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import React, { useState, useEffect, useRef } from 'react';
+import { USER_ID } from './api/todos';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
+import { client } from './utils/fetchClient';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Footer } from './components/Footer';
+import { TodoList } from './components/TodoList';
+import { Header } from './components/Header';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<Filter>('all');
+  const [error, setError] = useState<string | null>(null);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [input, setInput] = useState('');
+  const [deletingTodoIds, setDeletingTodoIds] = useState<number[]>([]);
+
+  const errorTimeout = useRef<number | null>(null);
+
+  function hideErrorButton() {
+    setError(null);
+  }
+
+  function handleNewFilterMode(mode: Filter) {
+    setFilter(mode);
+  }
+
+  function showError(message: string) {
+    if (errorTimeout.current !== null) {
+      clearTimeout(errorTimeout.current);
+    }
+
+    setError(message);
+    errorTimeout.current = window.setTimeout(() => setError(null), 3000);
+  }
+
+  function onInputChange(value: string) {
+    setInput(value);
+  }
+
+  const deleteTodo = async (id: number) => {
+    setDeletingTodoIds(prev => [...prev, id]);
+
+    try {
+      await client.delete(`/todos/${id}`);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+    } catch {
+      showError('Unable to delete a todo');
+    } finally {
+      setDeletingTodoIds(prev => prev.filter(todoId => todoId !== id));
+    }
+  };
+
+  async function handleClearCompleted() {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    await Promise.all(completedTodos.map(todo => deleteTodo(todo.id)));
+  }
+
+  function handleNewTitle(title: string) {
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: title,
+      completed: false,
+    };
+
+    setTempTodo(newTodo);
+
+    const addTodo = async () => {
+      try {
+        setIsLoading(true);
+        const response = await client.post<Todo>(`/todos`, newTodo);
+
+        setTodos(prevTodos => [...prevTodos, response]);
+        setTempTodo(null);
+        setInput('');
+      } catch {
+        showError('Unable to add a todo');
+        setTempTodo(null);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    addTodo();
```

Comment:
```suggestion
      try {
        setIsLoading(true);
        const response = await client.post<Todo>(`/todos`, newTodo);

        setTodos(prevTodos => [...prevTodos, response]);
        setTempTodo(null);
        setInput('');
      } catch {
        showError('Unable to add a todo');
        setTempTodo(null);
      } finally {
        setIsLoading(false);
      }
```

looks like redundant create addTodo function and immediately call it in one place

Reply:
@vitalii-bondarenko-ua Hi! Thanks for the review! Moved input and onInputChange to Header component, also simplified addTodo part. But I left handleNewTitle in App, because it also controls other states (Todos, TempTodo, isLoading)